### PR TITLE
fix(geosuggest): add Places API sessionToken support (issue #394)

### PR DESCRIPTION
**Pull request by KovDimaY**
### Description

Take a look at the issue #394.
  
As it is described in the [documentation of Google ](https://developers.google.com/maps/documentation/javascript/places-autocomplete), to add the session token support for `AutocompleteService`, it is just needed to add it to the request:
image

so this is shat I have done in this PR to enable session token support for this widget.

### Checklist

- [x] All tests passing
- [x] Created tests which fail without the change (if possible)
- [x] Extended the README / documentation, if necessary
- [x] Commits and PR follow conventions

### CODE REVIEW

**Comment by KovDimaY** [comment]
Hi, @ro-ka! I hope you have seen this PR. Let me know what do you think! Thank you! :)

**Comment by mikemonaco** [comment]
I _think_ this solution may be incomplete. A sessionToken can only be used for a sequence of predictions, then a single place details call. After that, a new session token must be generated.

Example of a  "session": keystroke, keystroke, keystroke, select place.

The session ends once you get the Place Details. If you continue to use that same token, you risk being charged per request. If you were to start typing again, a new sessionToken would be appropriate. 

"Place Autocomplete can use session tokens to group together autocomplete requests for billing purposes. A session token is a random string that is used to uniquely identify a session. A session consists of the activities required to resolve user input to a place. When a session token is passed (using the optional session_token parameter), autocomplete requests are not billed independently, but are instead billed once after a full autocomplete result is returned."

"Warning: Be sure to pass a unique session token for each new session. Using the same token for more than one session will result in each request being billed individually."

https://developers.google.com/places/web-service/autocomplete#session_tokens

**Comment by KovDimaY** [comment]
Thank you a lot, @mikemonaco!!! Thanks to your comment I made a deeper research and have found the next info about sessions and autocomplete:
image

The thing is that in this widget we are not doing any Place Details but we get info using Geocoder, so as I understand from the documentation the sessionToken can be the same (until it expires). But then, taking also a deeper look at the pricint, I have found that calling Autocomplete per session + Place Details = 0$ + 17$, but calling Autocomplete per session + Geocoding  (without Place Details) = 17$ + 5$. 
I got it based on this info:
image

image

So it is clear that much better it will be to change Geocoder to PlacesService and to save 5$ per 1000 requests. Anyway we are not using anything except lat and lng from the request result...

I will update the PR and wat for your opinion! 

Thank you for the contribution! :D

**Comment by ro-ka** [comment]
Thanks for bringing this topic up! We should definitely save on requests with the component.

The *Geocoder* can geocode a fixture which is not necessarily a place defined in Google Maps. But can the *PlacesService* do so, too?

**Comment by KovDimaY** [comment]
Hi, @ro-ka! Thank you for the comments! Do you want me to fix the conflicts and update the branch or you want to do it yourself? Let me know what should I do from my side! :D 
Thank you!

File: src/Geosuggest.jsx [hunk 1]
```diff
@@ -417,9 +422,9 @@ class Geosuggest extends React.Component {
           : null
       };
     }
-    this.geocoder.geocode(options, (results, status) => {
-      if (status === this.googleMaps.GeocoderStatus.OK) {
-        var gmaps = results[0],
+    this.placesService.getDetails(options, (results, status) => {
```
[the remark this answers is not in the document]
Comment (KovDimaY) [reply src/Geosuggest.jsx]: Yes, it definitely makes sense! I will change it! Thank you for the point!

**Comment by KovDimaY** [comment]
To be honest, this change to TypeSctript makes me struggle a lot. I never worked before whit the TS and I put a lot of effort in order to be able to run my fixes without compilation errors. But when I finally could run the project to check if my changes are working correctly, I got **RefererNotAllowedMapError** error for the http://0.0.0.0:8000... 
I feel a bit confused facing so many troubles for a relatively easy fix. I will spend a bit more trying to get it working, but I am quite close to quitting... :(

**Comment by ro-ka** [comment]
Hi! I'm sorry to hear that. The switch to TypeScript was made to make the component safer regarding the data passed around. Found some inconsistencies in the refactoring.

The error you’re getting has nothing to with TypeScript — that’s the Google Maps that requires that the domains are whitelisted, see [here](https://stackoverflow.com/questions/35288250/google-maps-javascript-api-referernotallowedmaperror). I’ll contact our Sales for this issue as thy manage the Maps keys.

I would apply the changes, too, but I don’t have access to the branch as it is in your profile. Hope you don’t quit and I’m really sorry for the not perfect setup!

**Comment by KovDimaY** [comment]
No no, @ro-ka, it is not your fault! It is my bad that I am not able yet to work productively with TS. Do not worry about it! I just was disappointed that I cannot have expected progress with the fix.

Yes, I know about the error from google maps, my point was that I had to create my own key with all the billing setup in order to continue testing if my fix is ok or not. I just did not want to provide to google my billing info, but also I understand why you have restricted this access for your key, because in theory if 1000 people will start contributing to the project and will test many times the component in local you can get a quite huge bill :D
Now I have my working key so it is not a problem anymore.

Well, I will try to do my best this weekend in order to get it working... Let's keep positive and optimistic. I hope, finally, I will nail it! 💪

**Comment by cybot1711** [comment]
> No no, @ro-ka, it is not your fault! It is my bad that I am not able yet to work productively with TS. Do not worry about it! I just was disappointed that I cannot have expected progress with the fix.
> 
> Yes, I know about the error from google maps, my point was that I had to create my own key with all the billing setup in order to continue testing if my fix is ok or not. I just did not want to provide to google my billing info, but also I understand why you have restricted this access for your key, because in theory if 1000 people will start contributing to the project and will test many times the component in local you can get a quite huge bill :D
> Now I have my working key so it is not a problem anymore.
> 
> Well, I will try to do my best this weekend in order to get it working... Let's keep positive and optimistic. I hope, finally, I will nail it!

Hey guys. I need a fix for this ASAP. Anything I can help with to get his out?

**Comment by KovDimaY** [comment]
I cannot do anything as I even cannot get it working on local... 
I clone the current version, on the master branch I run ` npm install`, then `npm start` and I get this error:
image

When I was trying to fix conflicts and apply my changes at the same time, I was getting a sequence of errors, fixing them step by step I got the one that I mentioned above and I could not move forward. I was thinking that these are my changes that cause this error so I tried to do everything from scratch, form a clean branch. And now when I clone this project from scratch I am getting this error every time and nothing works, the app is completely broken...

Some ideas why this can happen? Node versions, npm versions, dependencies versions? I need to start at least from some working version in order to have some progress... :(

**Comment by KovDimaY** [comment]
Some suggestion is that maybe it happens because almost any of the dependencies are fixed so some stupid dependency was updated with a breaking change and now nothing works. Why do not you fix the versions in package.json?

**Comment by cybot1711** [comment]
@KovDimaY I am having the same issues. There is a problem with rollup, its not finding the named exports form react. @ro-ka Are you aware of this?

**Comment by cybot1711** [comment]
@KovDimaY add the following to your example config, `npm start` uses example config.
```
import typescript from 'rollup-plugin-typescript2';
import commonjs from 'rollup-plugin-commonjs';
import resolve from 'rollup-plugin-node-resolve';
import replace from 'rollup-plugin-replace';
import {uglify} from 'rollup-plugin-uglify';

const env = process.env.NODE_ENV;

export default {
  input: 'example/src/app.tsx',
 
  output: {
    name: 'Geosuggest',
    sourcemap: env === 'production' ? false : 'inline',
    exports: 'none',
    globals: ['google'],
    format: 'iife',
    file: 'example/dist/app.js'
  },
  plugins: [
    typescript(),
    resolve({jsnext: true, main: true, browser: true}),
    commonjs({
      include: 'node_modules/**',
      namedExports: {
        'react': [ 'Children',
        'Component',
        'PureComponent',
        'PropTypes',
        'createElement',
        'Fragment',
        'cloneElement',
        'StrictMode',
        'createFactory',
        'createRef',
        'createContext',
        'isValidElement',
        'isValidElementType',]
      }
    }),
    replace({'process.env.NODE_ENV': JSON.stringify(env)}),
    env === 'production' && uglify()
  ],
};
```

**Comment by KovDimaY** [comment]
Thank you, @cybot1711! Your solution worked out. 👍 
I have updated my PR and checked it on my local. Looks like everything works fine. Let's wait for the owners' review.

**Comment by cybot1711** [comment]
@ro-ka Can we please have a review on this?

**Comment by ro-ka** [comment]
Looks very good, thanks for not giving up!  🎉 ❤️

**Comment by cybot1711** [comment]
@ro-ka Don't forget to update npm package.

**Comment by KovDimaY** [comment]
Thank you, @ro-ka! It was a great challenge for me! 💪 
Hope the fix will help many people! 😇

**Comment by ro-ka** [comment]
Published as `2.9.0`.  Thanks everybody!  🎉

**Pull request merged**